File: benchmark-java/src/androidTest/java/com/shinkevich/benchmark_java/AsyncDBBenchmark.java
package com.shinkevich.benchmark_java;

import androidx.benchmark.BenchmarkState;
import androidx.benchmark.junit4.BenchmarkRule;
import androidx.lifecycle.Observer;
import androidx.lifecycle.testing.TestLifecycleOwner;
import androidx.test.platform.app.InstrumentationRegistry;

import com.shinkevich.benchmarkable.DBTest;
import com.shinkevich.benchmarkable.Person;
import com.shinkevich.benchmarkable.PersonDao;

import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.schedulers.Schedulers;

@RunWith(value = Parameterized.class)
public class AsyncDBBenchmark {
    private static final String TAG = AsyncDBBenchmark.class.getSimpleName();
    private static final Integer ALPHABET_SIZE = 26;

    @Rule
    public BenchmarkRule benchmarkRule = new BenchmarkRule();
    @Parameterized.Parameter
    public int tasksNum;

    @Parameterized.Parameters
    public static Collection<Object[]> tasksNumValues() {
        return Arrays.asList(new Object[][]{
                {5}, {10}, {15}, {20}, {25}, {30}, {35}, {40}, {45}, {50}});
    }

    @Test
    public void rxJavaSingleTest() {
        BenchmarkState state = benchmarkRule.getState();
        DBTest dbTest = new DBTest(InstrumentationRegistry.getInstrumentation().getContext());
        PersonDao personDao = dbTest.getPersonDao();
        while (state.keepRunning()) {
            benchmarkRule.getScope().runWithTimingDisabled(() -> {
                dbTest.initDatabase();
                return null;
            });
            Observable
                    .range(0, tasksNum)
                    .flatMapSingle(taskNum -> personDao.getByNameStartSingle(getNameStartForSearch(taskNum)))
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .blockingSubscribe();
        }
    }

    @Test
    public void rxJavaObservableTest() {
        BenchmarkState state = benchmarkRule.getState();
        DBTest dbTest = new DBTest(InstrumentationRegistry.getInstrumentation().getContext());
        PersonDao personDao = dbTest.getPersonDao();
        while (state.keepRunning()) {
            benchmarkRule.getScope().runWithTimingDisabled(() -> {
                dbTest.initDatabase();
                return null;
            });
            Observable
                    .range(0, tasksNum)
                    .flatMap(taskNum -> personDao.getByNameStartObservable(getNameStartForSearch(taskNum)).take(1))
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .blockingSubscribe();
        }
    }

    @Test
    public void executorTest() throws ExecutionException, InterruptedException {
        BenchmarkState state = benchmarkRule.getState();
        DBTest dbTest = new DBTest(InstrumentationRegistry.getInstrumentation().getContext());
        PersonDao personDao = dbTest.getPersonDao();
        ExecutorService executorService = Executors.newCachedThreadPool();
        while (state.keepRunning()) {
            benchmarkRule.getScope().runWithTimingDisabled(() -> {
                dbTest.initDatabase();
                return null;
            });
            List<Future<List<Person>>> results = new ArrayList<>();
            for (int i = 0; i < tasksNum; i++) {
                int finalI = i;
                results.add(executorService.submit(() -> personDao.getByNameStart(getNameStartForSearch(finalI))));
            }
            for (Future<List<Person>> result : results) {
                result.get();
            }
        }
    }

//    @Test
//    public void liveDataTest() throws InterruptedException {
//        BenchmarkState state = benchmarkRule.getState();
//        DBTest dbTest = new DBTest(InstrumentationRegistry.getInstrumentation().getContext());
//        PersonDao personDao = dbTest.getPersonDao();
//        CountDownLatch countDownLatch = new CountDownLatch(tasksNum);
//        while (state.keepRunning()) {
//            benchmarkRule.getScope().runWithTimingDisabled(() -> {
//                dbTest.initDatabase();
//                return null;
//            });
//            for (int i = 0; i < tasksNum; i++) {
//                personDao.getByNameStartLiveData("a").observe(new TestLifecycleOwner(), new Observer<List<Person>>() {
//                    @Override
//                    public void onChanged(List<Person> people) {
//                        countDownLatch.countDown();
//                    }
//                });
//            }
//            countDownLatch.await();
//        }
//    }

    private String getNameStartForSearch(Integer taskNum) {
        return Character.toString((char) (taskNum % ALPHABET_SIZE + 'a'));
    }
}

File: benchmark-kotlin/src/androidTest/java/com/shinkevich/benchmark_kotlin/AsyncDBBenchmark.kt
package com.shinkevich.benchmark_kotlin

import androidx.benchmark.junit4.BenchmarkRule
import androidx.benchmark.junit4.measureRepeated
import androidx.test.platform.app.InstrumentationRegistry
import com.shinkevich.benchmarkable.DBTest
import com.shinkevich.benchmarkable.Person
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.asFlow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.flatMapMerge
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.flow.merge
import kotlinx.coroutines.flow.take
import kotlinx.coroutines.runBlocking
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized

@RunWith(value = Parameterized::class)
class AsyncDBBenchmark(
    private val tasksNum: Int
) {
    @get:Rule
    var benchmarkRule = BenchmarkRule()

    @Test
    fun suspendTest() {
        val dbTest = DBTest(InstrumentationRegistry.getInstrumentation().context)
        val personDao = dbTest.personDao
        benchmarkRule.measureRepeated {
            runWithTimingDisabled { dbTest.initDatabase() }
            val results = mutableListOf<Deferred<List<Person>>>()
            runBlocking(Dispatchers.Main) {
                repeat(tasksNum) {
                    results.add(CoroutineScope(Dispatchers.IO).async {
                        personDao.getByNameStartSuspendable(getNameStartForSearch(it))
                    })
                }
                results.awaitAll()
            }
        }
    }

    @Test
    fun flowTest() {
        val dbTest = DBTest(InstrumentationRegistry.getInstrumentation().context)
        val personDao = dbTest.personDao
        benchmarkRule.measureRepeated {
            runWithTimingDisabled { dbTest.initDatabase() }
            runBlocking(Dispatchers.Main) {
                (0 until tasksNum).asFlow()
                    .flatMapMerge { taskNum ->
                        personDao.getByNameStartFlow(getNameStartForSearch(taskNum)).take(1)
                    }
                    .flowOn(Dispatchers.IO)
                    .collect()
            }
        }
    }

    @Test
    fun flowTest2() {
        val dbTest = DBTest(InstrumentationRegistry.getInstrumentation().context)
        val personDao = dbTest.personDao
        benchmarkRule.measureRepeated {
            runWithTimingDisabled { dbTest.initDatabase() }
            runBlocking(Dispatchers.Main) {
                val results = mutableListOf<Flow<List<Person>>>()
                for(taskNum in 0 until tasksNum){
                    results.add(personDao.getByNameStartFlow(getNameStartForSearch(taskNum)))
                }
                results
                    .merge()
                    .take(tasksNum)
                    .flowOn(Dispatchers.IO)
                    .collect()
            }
        }
    }

    private fun getNameStartForSearch(taskNum: Int): String {
        return (taskNum % ALPHABET_SIZE + 'a'.code).toChar().toString()
    }

    companion object {
        private val TAG = AsyncDBBenchmark::class.java.simpleName
        private const val ALPHABET_SIZE = 26

        @JvmStatic
        @Parameterized.Parameters
        fun data(): Iterable<Array<Any>> {
            return arrayListOf(
                arrayOf(5),
                arrayOf(10),
                arrayOf(15),
                arrayOf(20),
                arrayOf(25),
                arrayOf(30),
                arrayOf(35),
                arrayOf(40),
                arrayOf(45),
                arrayOf(50)
            ).toList() as Iterable<Array<Any>>
        }
    }
}
